pac4j-sql/src/main/java/org/pac4j/sql/credentials/authenticator/DbAuthenticator.java

```diff
@@ -0,0 +1,173 @@
+        this.attributes = attributes;
+    }
+
+    public DbAuthenticator(final DataSource dataSource, final String attributes, final PasswordEncoder passwordEncoder) {
+        this.dataSource = dataSource;
+        this.attributes = attributes;
+        this.passwordEncoder = passwordEncoder;
+    }
+
+    protected void initDbi() {
+        if (this.dbi == null) {
+            synchronized (this) {
+                if (this.dbi == null) {
+                    this.dbi = new DBI(this.dataSource);
+                }
+            }
+        }
+    }
+
+    public void validate(UsernamePasswordCredentials credentials) {
+        CommonHelper.assertNotNull("dataSource", this.dataSource);
+        CommonHelper.assertNotNull("attributes", this.attributes);
+        CommonHelper.assertNotNull("passwordEncoder", this.passwordEncoder);
+
+        initDbi();
+
+        Handle h = null;
+        try {
+            h = dbi.open();
+
+            final String username = credentials.getUsername();
+            final String query;
+            if (CommonHelper.isNotBlank(attributes)) {
+                query = startQuery + ", " + attributes + endQuery;
+            } else {
+                query = startQuery + endQuery;
+            }
+            final List<Map<String, Object>> results = h.createQuery(query).bind("username", username).list(2);
+
+            if (results == null || results.size() == 0) {
+                throw new AccountNotFoundException("No account found for: " + username);
+            } else if (results.size() > 1) {
+                throw new MultipleAccountsFoundException("Too many accounts found for: " + username);
+            } else {
+                final Map<String, Object> result = results.get(0);
+                final String expectedPassword = passwordEncoder.encode(credentials.getPassword());
+                final String returnedPassword = (String) result.get("password");
+                if (CommonHelper.areNotEquals(returnedPassword, expectedPassword)) {
+                    throw new BadCredentialsException("Bad credentials for: " + username);
+                } else {
+                    final DbProfile profile = createProfile(username, attributes.split(","), result);
+                    credentials.setUserProfile(profile);
+                }
+            }
+
+        } catch (final TechnicalException e) {
+            logger.error("Authentication error", e);
+            throw e;
+        } catch (final RuntimeException e) {
+            logger.error("Cannot fetch username / password from DB", e);
+            throw new TechnicalException(e);
+        } finally {
+            if (h != null) {
+                h.close();
+            }
+        }
+    }
+
+    protected DbProfile createProfile(final String username, final String[] attributes, final Map<String, Object> result) {
+        final DbProfile profile = new DbProfile();
+        profile.setId(username);
+        for (String attribute: attributes) {
```

Comment:
You mean loop on the `result` map and populate the `profile`. It will be the same result, won't it? What do you have in mind with "work better"?